code/datums/spell_cooldown/spell_cooldown.dm

```diff
@@ -0,0 +1,57 @@
+/datum/spell_cooldown
+	/// the world.time the spell will be available again
+	var/recharge_time = 0
+	/// the amount of time that must pass before a spell can be used again
+	var/recharge_duration = 10 SECONDS // default spell cooldown
+	/// does it start off cooldown?
+	var/starts_off_cooldown = TRUE
+	/// holds a ref to the spell
+	var/obj/effect/proc_holder/spell/spell_parent = null
+
+/datum/spell_cooldown/Destroy()
+	spell_parent = null
+	return ..()
+
+/datum/spell_cooldown/proc/cooldown_init(obj/effect/proc_holder/spell/new_spell)
+	spell_parent = new_spell
+	if(!starts_off_cooldown)
+		start_recharge()
+
+/datum/spell_cooldown/proc/is_on_cooldown()
+	return recharge_time > world.time
+
+/datum/spell_cooldown/proc/should_end_cooldown()
+	return !is_on_cooldown()
+
+/datum/spell_cooldown/proc/end_recharge()
+	return
+
+/datum/spell_cooldown/process()
+	if(spell_parent.action)
+		spell_parent.action.UpdateButtonIcon()
```

Comment:
Can this be null at one point? Maybe due to a bug I guess.
And should it continue processing when the action is null?